// @ridleywinters/sea/src/parsers/wikilinks.js
import * as base from '../base';

// Forked from https://github.com/djm/remark-shortcodes
/*
	 remark-shortcodes

	 A remark plugin to allow custom block-level Wordpress/Hugo-like
	 shortcodes to be parsed as part of the Remark Markdown AST. Note
	 that this plugin does not apply any transformations, but simply
	 parses the shortcodes and adds the relevant data to the AST. The
	 start & end blocks are configurable, but the inner shortcode
	 including the name and the key="value" pairs must be as expected.

	 To apply transformations from the shortcodes to HTML or other
	 formats, please see the example in the README.

	 e.g
				// with no attributes
				[[ ShortcodeNameA ]]

				// with multiple attributes
				[[ ShortcodeNameC a="b" c="d" ]]

				// with custom start/end Blocks
				{{% ShortcodeNameD a="1" c="2" %}}

				// An example node format for a shortcode with attributes
				{
					type: 'shortcode',
					identifier: 'ShortcodeNameC',
					attributes: {
							a: '1',
							c: '2'
					}
				}

	 With thanks to official remark plugins:
	 https://github.com/wooorm/remark-breaks/blob/master/index.js
	 https://github.com/wooorm/remark-gemoji/blob/master/index.js
*/

module.exports = wikilinks;

function wikilinks(options) {
	var startBlock = (options || {}).startBlock || "[[";
	var endBlock = (options || {}).endBlock || "]]";
	var inlineMode = (options || {}).inlineMode || false;

	if (isRemarkParser(this.Parser)) {
		var parser = this.Parser.prototype;
		var tokenizers = inlineMode ? parser.inlineTokenizers : parser.blockTokenizers;
		var methods = inlineMode ? parser.inlineMethods : parser.blockMethods;

		tokenizers.wikilink = shortcodeTokenizer;
		methods.splice(methods.indexOf("html"), 0, "wikilink");
	}
	if (isRemarkCompiler(this.Compiler)) {
		var compiler = this.Compiler.prototype;
		compiler.visitors.wikilink = shortcodeCompiler;
	}

	function locator(value, fromIndex) {
		return value.indexOf(startBlock, fromIndex);
	}

	function shortcodeTokenizer(eat, value, silent) {
		var entireShortcode;
		var innerShortcode;
		var parsedShortcode;
		var endPosition;
		var endBlockPosition;

		if (!value.startsWith(startBlock)) return;

		endBlockPosition = value.indexOf(endBlock, startBlock.length);
		if (endBlockPosition === -1) return;

		endPosition = endBlockPosition + endBlock.length;
		entireShortcode = value.slice(0, endPosition);
		innerShortcode = value.slice(startBlock.length, endBlockPosition);

		parsedShortcode = parseShortcode(innerShortcode);

		// If there is no parsed data, something fishy is up - return nothing.
		if (!parsedShortcode) return;

		/* Exit with true in silent mode after successful parse - never used (yet) */
		/* istanbul ignore if */
		if (silent) {
			return true;
		}
		return eat(entireShortcode)(parsedShortcode);
	}
	shortcodeTokenizer.locator = locator;

	function shortcodeCompiler(node) {
		var attributes = "";
		var keys = Object.keys(node.attributes || {});
		if (keys.length > 0) {
			attributes =
				" " +
				keys
					.map(function (key) {
						return key + '="' + node.attributes[key] + '"';
					})
					.join(" ");
		}
		return startBlock + " " + node.identifier + attributes + " " + endBlock;
	}
}

/**
 * Parses the contents of the raw text within the wikilink.
 */
function parseShortcode(text) {
	const node = {
		type: "wikilink",
		id: null,
		text: null,
	}

	let parts = text.split('|');
	

	if (parts.length === 2) {
		node.text = parts[0];
		node.id = base.normalizeID(parts[1]);
	} else {
		node.id = base.normalizeID(parts[0]);
	}
	return node;
}

function isRemarkParser(parser) {
	return Boolean(
		parser &&
		parser.prototype &&
		parser.prototype.inlineTokenizers &&
		parser.prototype.inlineTokenizers.break &&
		parser.prototype.inlineTokenizers.break.locator
	);
}

function isRemarkCompiler(compiler) {
	return Boolean(compiler && compiler.prototype);
}
